### PR TITLE
Add VirtIO Viomem Driver support

### PR DESCRIPTION
**Pull request by kostyanf14**
@YanVugenfirer Please wait for merge untill get answer https://github.com/virtio-win/kvm-guest-drivers-windows/pull/794#issuecomment-1192287484

### CODE REVIEW

File: lib/setupmanagers/qemuhck/qemu_machine.rb [hunk 1]
```diff
@@ -193,6 +193,13 @@ def options_replacement_list
       }
     end
 
+    def double_memory
```
Comment (YanVugenfirer) [inline lib/setupmanagers/qemuhck/qemu_machine.rb:196]: Seems confusing name for function and variable. 
Maybe: memory_for_virtio_mem or pluggable_memory?

Comment (kostyanf14) [reply lib/setupmanagers/qemuhck/qemu_machine.rb:196]: No, this is just double memory. We run QEMU with option -m memory,maxmemory=double_memory. I hope that this reserve will be enough. While memory_for_virtio_mem is 1G in our configuration and pluggable_memory is the same as a run memory.

**Comment by kostyanf14** [comment]
PR was rebased to current master

**Review by akihikodaki — COMMENTED** [review]
Why do you need `double_memory`?

**Comment by akihikodaki** [comment]
> Why do you need `double_memory`?

Nevermind. I read the help message of QEMU.

**Review by akihikodaki — CHANGES_REQUESTED** [review]
That parsing is a bit tricky. Why don't you just convert `memory` into a normal JSON number so that we don't need parsing?

File: lib/setupmanagers/qemuhck/devices/virtio-mem-pci.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "name": "virtio-mem-pci",
+  "memory_extra_param": ",maxmem=@double_memory@",
```
Comment (akihikodaki) [inline lib/setupmanagers/qemuhck/devices/virtio-mem-pci.json:3]: Why don't you always add `maxmem` and instead make it specific to virtio-mem-pci?

Comment (kostyanf14) [reply lib/setupmanagers/qemuhck/devices/virtio-mem-pci.json:3]: In the case of NetKVM (3 VM), we will have the sum of maxmem greater than the host RAM on CI servers. I want to be sure that the host always has some free memory.

**Comment by kostyanf14** [comment]
>That parsing is a bit tricky. Why don't you just convert memory into a normal JSON number so that we don't need parsing?

Do you suggest to convert 4G into 4294967296?

File: lib/setupmanagers/qemuhck/devices/virtio-mem-pci.json (hunk 2 above)
Comment (akihikodaki) [reply lib/setupmanagers/qemuhck/devices/virtio-mem-pci.json:3]: Is it problematic?

**Comment by akihikodaki** [comment]
> > That parsing is a bit tricky. Why don't you just convert memory into a normal JSON number so that we don't need parsing?
> 
> Do you suggest to convert 4G into 4294967296?

No, just 4.

**Comment by kostyanf14** [comment]
> > > That parsing is a bit tricky. Why don't you just convert memory into a normal JSON number so that we don't need parsing?
> > 
> > 
> > Do you suggest to convert 4G into 4294967296?
> 
> No, just 4.

And suppose that we always use GB as a unit? Maybe rename key name to `memory_gb` in this case?

**Comment by akihikodaki** [comment]
> > > > That parsing is a bit tricky. Why don't you just convert memory into a normal JSON number so that we don't need parsing?
> > > 
> > > 
> > > Do you suggest to convert 4G into 4294967296?
> > 
> > 
> > No, just 4.
> 
> And suppose that we always use GB as a unit? Maybe rename key name to `memory_gb` in this case?

Yes. Renaming key is a nice suggestion.

File: lib/setupmanagers/qemuhck/devices/virtio-mem-pci.json (hunk 2 above)
Comment (akihikodaki) [reply lib/setupmanagers/qemuhck/devices/virtio-mem-pci.json:3]: I think it's not that strange that the sum of maxmem is greater than the host RAM.

Think of a multi-tenant machine farm; there are multiple machines, and each machine hosts multiple VMs. Also, assume that you want to scale VMs bigger without downtime.
In such a scenario, you set `maxmem` of all machines to match the maximum memory size a VM can hold. The sum of `maxmem` of VMs in a machine can surpass the physical memory size. Later, if too many memory is requested for a machine, some VMs on the machine will migrate to make space.
I think some cloud vendors already do this kind of trick.

Comment (kostyanf14) [reply lib/setupmanagers/qemuhck/devices/virtio-mem-pci.json:3]: I don't know the QEMU source related to `maxmem` option. If you can confirm that QEMU will not allocate more than memory_size on Windows Guest without user action. I am ok.

Comment (akihikodaki) [reply lib/setupmanagers/qemuhck/devices/virtio-mem-pci.json:3]: You can assume that; I believe the very purpose of `maxmem` is to support a scenario I have just described, and in that case, we should never have a problem with overcomitted `maxmem`.

Comment (kostyanf14) [reply lib/setupmanagers/qemuhck/devices/virtio-mem-pci.json:3]: Ok. Let's assume that. We can always change this if see any problem.

File: lib/setupmanagers/qemuhck/qemu_machine.rb [hunk 3]
```diff
@@ -358,6 +358,7 @@ def full_replacement_list
         '@client_id@' => @client_id,
         '@workspace@' => @workspace_path,
         '@memory@' => "#{option_config('memory_gb')}G",
+        '@double_memory@' => "#{2 * option_config('memory_gb')}G",
```
Comment (akihikodaki) [inline lib/setupmanagers/qemuhck/qemu_machine.rb:361]: Let's have `@pluggable_memory@` and `@max_memory@`; `@max_memory` will be the sum of `@memory@` and `@pluggable_memory@`.

**Review by akihikodaki — CHANGES_REQUESTED** [review]

File: lib/setupmanagers/qemuhck/qemu_machine.rb (hunk 3 above)
Comment (kostyanf14) [reply lib/setupmanagers/qemuhck/qemu_machine.rb:361]: In this case, I need to talk with the driver developer. You can see that memory-backend-ram and virtio-mem-pci have a different memory size. What is `@pluggable_memory@` now `2G` or `1G`? That's why I don't prefer ideas with calculation.

Comment (akihikodaki) [reply lib/setupmanagers/qemuhck/qemu_machine.rb:361]: The same applies even now; changing the size of `mem_backend` into 512M will certainly cause a trouble. Having `@pluggable_memory@` only moves the place where the value `2G` is written, and allows calculating `@max_memory@`.

That said, I see no reason that `request-size` has to be inequal to `mem_backend` size.

Comment (kostyanf14) [reply lib/setupmanagers/qemuhck/qemu_machine.rb:361]: Just checked QE command line `extra_cmd= ' -object memory-backend-ram,id=vmem0,size=5G -device virtio-mem-pci,id=mem0,memdev=vmem0,node=0,requested-size=2G,indirect_desc=off,bus='+bus_num `

`request-size` != `mem_backend` size.

Comment (akihikodaki) [reply lib/setupmanagers/qemuhck/qemu_machine.rb:361]: It's better to ask if changing `requested-size` is OK.

**Comment by kostyanf14** [comment]
@akihikodaki 
Let's continue with request-size != mem_backend size.

What comments do you have in this case?

**Comment by akihikodaki** [comment]
> @akihikodaki Let's continue with request-size != mem_backend size.
> 
> What comments do you have in this case?

The comment in https://github.com/HCK-CI/AutoHCK/pull/143/files#r1495953291 still applies. I have nothing else to add.

**Comment by akihikodaki** [comment]
> > > @akihikodaki Let's continue with request-size != mem_backend size.
> > > What comments do you have in this case?
> > 
> > 
> > The comment in https://github.com/HCK-CI/AutoHCK/pull/143/files#r1495953291 still applies. I have nothing else to add.
> 
> How do you expect to see the syntax of the @max_memory@ calculation? Or do you expect this as the internal logic of AutoHCK if @pluggable_memory@ is defined?

I expect an internal logic.

**Review by akihikodaki — CHANGES_REQUESTED** [review]
Why don't you make `pluggable_memory_gb` a number defined with platform as `memory_gb` is?

**Comment by kostyanf14** [comment]
>Why don't you make pluggable_memory_gb a number defined with platform as memory_gb is?

I am ok with making pluggable_memory_gb a number, but it doesn't relate to a platform. 
This fully depends on a device for now.

**Review by akihikodaki — CHANGES_REQUESTED** [review]

**Comment by akihikodaki** [comment]
> > Why don't you make pluggable_memory_gb a number defined with platform as memory_gb is?
> 
> I am ok with making pluggable_memory_gb a number, but it doesn't relate to a platform. This fully depends on a device for now.

I don't think `pluggable_memory_gb` is that different from `memory_gb`. The platform provides memory, and the devices and tests use it. Tests would just fail if the provided memory is not sufficient. We can use the same logic for pluggable memory as well. But It's not a strong opinion and I won't insist.

In any case, don't put it in `define_variables`. It is actually a list of string replacements and better to be typed as `T::Hash[String, String]`. Instead, put `pluggable_memory_gb` at the root of the device JSON.

File: lib/setupmanagers/qemuhck/qemu_machine.rb [hunk 4]
```diff
@@ -402,6 +413,9 @@ def load_device_info(device_info)
           var_value.merge! value
         when Array
           var_value << value
+        when Integer, String
+          @logger.info("Replacing #{var} value #{var_value} -> #{value}")
```
Comment (akihikodaki) [inline lib/setupmanagers/qemuhck/qemu_machine.rb:417]: Is this log really needed? We don't emit logs when overwriting values for `Hash`.

**Review by akihikodaki — CHANGES_REQUESTED** [review]

File: lib/setupmanagers/qemuhck/qemu_machine.rb (hunk 4 above)
Comment (kostyanf14) [reply lib/setupmanagers/qemuhck/qemu_machine.rb]: removed

**Review by akihikodaki — APPROVED** [review]

**Pull request merged**